### PR TITLE
[OP] add BilinearSamplingOp and GridGeneratorOp

### PR DESCRIPTION
**Pull request by sxjscience**
Continuing https://github.com/dmlc/mxnet/pull/4728

### CODE REVIEW

**Comment by sxjscience** [comment]
@dsqx71 @piiswrong

File: src/operator/bilinear_sampler.cu [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+                                    const Tensor<gpu, 4, DType> &input,
+                                    const Tensor<gpu, 4, DType> &grid_src) {
+    DType *out = output.dptr_;
+    const DType *data = input.dptr_;
+    const DType *grid = grid_src.dptr_;
+    int o_n = output.size(0), o_c = output.size(1), o_h = output.size(2), o_w = output.size(3);
+    int i_c = input.size(1), i_h = input.size(2), i_w = input.size(3);
+    using namespace cuda;
+    const int max_block = (output.shape_.Size() + kMaxThreadsPerBlock - 1) / kMaxThreadsPerBlock;
+    dim3 num_blocks(kMaxGridNum, (max_block + kMaxGridNum - 1) / kMaxGridNum);
```
Comment (piiswrong) [inline src/operator/bilinear_sampler.cu]: are you sure this is right?

File: src/operator/bilinear_sampler.cc [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+}
+
+DMLC_REGISTER_PARAMETER(BilinearSamplerParam);
+
+MXNET_REGISTER_OP_PROPERTY(BilinearSampler, BilinearSamplerProp)
+.add_argument("data", "Symbol", "Input data to the BilinearsamplerOp.")
+.add_argument("grid", "Symbol", "Input grid to the BilinearsamplerOp."
+                                "grid has two channels: x_src, y_src")
+.add_arguments(BilinearSamplerParam::__FIELDS__())
+.describe("Apply bilinear sampling to input feature map.\n    "
```
Comment (piiswrong) [inline src/operator/bilinear_sampler.cc]: I still think some people will search for spatial transformer when looking for this layer.
Let's say "bilinear sampling (spatial transformer layer)" here and quote some relevent papers by title

File: src/operator/bilinear_sampler.cu [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+  int o_n = output_grad.size(0), o_c = output_grad.size(1),
+      o_h = output_grad.size(2), o_w = output_grad.size(3);
+  int i_c = input_data.size(1), i_h = input_data.size(2), i_w = input_data.size(3);
+  using namespace cuda;
+  const int max_block = (output_grad.shape_.Size() / o_c + kMaxThreadsPerBlock - 1)
+                        / kMaxThreadsPerBlock;
+  const int grid_dim_x = (max_block > kMaxGridNum) ? kMaxGridNum : max_block;
+  const int grid_dim_y =
+    (max_block > kMaxGridNum) ? (max_block + kMaxGridNum - 1) / kMaxGridNum : 1;
+  dim3 num_blocks(grid_dim_x, grid_dim_y);
```
Comment (sxjscience) [inline src/operator/bilinear_sampler.cu]: Changed to this. The old launching parameter was too large.

File: tests/jenkins/run_test.sh [hunk 4]
```diff
@@ -17,7 +17,7 @@ if [ $user == 'root' ]
 then
     python setup.py install || exit 1
 else
-    python setup.py install --prefix ~/.local || exit 1
+    python setup.py install --user || exit 1
```
Comment (piiswrong) [inline tests/jenkins/run_test.sh]: @lxn2 I over looked this. Why did you add the installation here?

@sxjscience Please remove them all together

File: tests/jenkins/run_test.sh [hunk 5]
```diff
@@ -8,38 +8,17 @@ echo "USE_CUDNN=1" >> config.mk
...
 
 echo "BUILD cpp_test"
 make -j$(nproc) test || exit -1
 export MXNET_ENGINE_INFO=true
-#for test in tests/cpp/*_test; do
-#    ./$test || exit -1
-#done
+for test in tests/cpp/*_test; do
+    ./$test || exit -1
+done
```
Comment (sxjscience) [inline tests/jenkins/run_test.sh]: I find that these cpp tests are removed. Should I add them back?

Comment (piiswrong) [inline tests/jenkins/run_test.sh]: yes

**Comment by piiswrong** [comment]
Looks like other installation of mxnet is interfering. add export PYTHONPATH=$(PWD)/python?

**Comment by piiswrong** [comment]
Please fix test then squash the commits so that there is one commit from Dong Xu and one commit from you

**Comment by piiswrong** [comment]
@sxjscience Could you make sure tests pass with USE_CUDNN=0? We currently don't run non-cudnn tests

**Comment by sxjscience** [comment]
@piiswrong I've tested locally that pure cuda works.

**Comment by sxjscience** [comment]
@piiswrong Would directly squash lose the authorship here?
